packages/network-controller/src/NetworkController.ts

```diff
@@ -1063,205 +1482,427 @@ export class NetworkController extends BaseController<
-   * @returns The ID for the added or updated network configuration.
+   * @param chainId - The chain ID associated with an existing network.
+   * @param fields - The object that describes the updates to the network/chain,
+   * including the new set of RPC endpoints which should front that chain.
+   * @param options - Options to provide.
+   * @param options.replacementSelectedRpcEndpointIndex - Usually you cannot
+   * remove an RPC endpoint that is being represented by the currently selected
+   * network client. This option allows you to specify another RPC endpoint
+   * (either an existing one or a new one) that should be used to select a new
+   * network instead.
+   * @returns The updated network configuration.
+   * @throws if `chainId` does not refer to an existing network configuration,
+   * if any part of `fields` would produce invalid state, etc.
+   * @see {@link NetworkConfiguration}
    */
-  async upsertNetworkConfiguration(
-    networkConfiguration: NetworkConfiguration,
+  async updateNetwork(
+    chainId: Hex,
+    fields: UpdateNetworkFields,
     {
-      referrer,
-      source,
-      setActive = false,
-    }: {
-      referrer: string;
-      source: string;
-      setActive?: boolean;
-    },
-  ): Promise<string> {
-    const sanitizedNetworkConfiguration: NetworkConfiguration = pick(
-      networkConfiguration,
-      ['rpcUrl', 'chainId', 'ticker', 'nickname', 'rpcPrefs'],
-    );
-    const { rpcUrl, chainId, ticker } = sanitizedNetworkConfiguration;
+      replacementSelectedRpcEndpointIndex,
+    }: { replacementSelectedRpcEndpointIndex?: number } = {},
+  ): Promise<NetworkConfiguration> {
+    const existingNetworkConfiguration =
+      this.state.networkConfigurationsByChainId[chainId];
 
-    assertIsStrictHexString(chainId);
-    if (!isSafeChainId(chainId)) {
+    if (existingNetworkConfiguration === undefined) {
       throw new Error(
-        `Invalid chain ID "${chainId}": numerical value greater than max safe value.`,
-      );
-    }
-    if (!rpcUrl) {
-      throw new Error(
-        'An rpcUrl is required to add or update network configuration',
-      );
-    }
-    if (!referrer || !source) {
-      throw new Error(
-        'referrer and source are required arguments for adding or updating a network configuration',
-      );
-    }
-    try {
-      new URL(rpcUrl);
-      // TODO: Replace `any` with type
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    } catch (e: any) {
-      if (e.message.includes('Invalid URL')) {
-        throw new Error('rpcUrl must be a valid URL');
-      }
-    }
-    if (!ticker) {
-      throw new Error(
-        'A ticker is required to add or update networkConfiguration',
+        `Could not update network: Cannot find network configuration for chain ${inspect(
+          chainId,
+        )}`,
       );
     }
 
+    const existingChainId = chainId;
+    const { chainId: newChainId, rpcEndpoints: setOfNewRpcEndpointFields } =
```

Comment:
does this mean that only `chainId` and `rpcEndpoints` are editable ? we're not extracting name and ticker from the `fields`

Reply:
All fields are editable. We destructure `chainId` and `rpcEndpoints` here so they're easier to work with on their own below, but the whole `fields` object ends up getting passed and updated in state, e.g.:

``` typescript
      this.update((state) => {
        this.#updateNetworkConfigurations({
          state,
          mode: 'update',
          networkFields: fields,
          networkConfigurationToPersist: updatedNetworkConfiguration,
          existingNetworkConfiguration,
        });
      });
```